subcommand/injector/command.go

```diff
@@ -316,82 +318,131 @@ func getAdminAPIVersion(ctx context.Context, clientset *kubernetes.Clientset) (s
 	return adminAPIVersion, err
 }
 
-func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, clientset *kubernetes.Clientset, leaderElector leader.Elector, adminAPIVersion string, log hclog.Logger) {
+func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, clientset *kubernetes.Clientset, leaderElector leader.Elector, log hclog.Logger) {
 	var bundle cert.Bundle
+	webhooksCache, webhooksWatcher, err := watchWebhooks(ctx, clientset)
+	if err != nil {
+		log.Error("Error watching webhooks", "error", err)
+		panic(err)
+	}
+
+	expBackoff := backoff.NewExponentialBackOff()
 
 	for {
 		select {
 		case bundle = <-ch:
-			log.Info("Updated certificate bundle received. Updating certs...")
 			// Bundle is updated, set it up
+			log.Info("Updated certificate bundle received. Updating certs...")
 
-		case <-time.After(1 * time.Second):
-			// This forces the mutating webhook config to remain updated
-			// fairly quickly. This is a jank way to do this and we should
-			// look to improve it in the future. Since we use Patch requests
-			// it is pretty cheap to do, though.
+		case <-webhooksWatcher:
+			// Webhooks changed, make sure the CA bundle is up-to-date.
+			log.Info("Webhooks changed. Updating certs...")
 
 		case <-ctx.Done():
 			// Quit
 			return
 		}
-
-		crt, err := tls.X509KeyPair(bundle.Cert, bundle.Key)
+		err := c.updateWebhook(ctx, clientset, bundle, webhooksCache, leaderElector, log)
 		if err != nil {
-			log.Warn(fmt.Sprintf("Could not load TLS keypair: %s. Trying again...", err))
-			continue
+			time.Sleep(expBackoff.NextBackOff())
```

Comment:
This will sleep for the backoff duration once, but it will block on the next for loop run until the next case is selected, which may be munch longer than the backoff duration here.

Reply:
Great point. We do want to retry here. I'll rework the loop.